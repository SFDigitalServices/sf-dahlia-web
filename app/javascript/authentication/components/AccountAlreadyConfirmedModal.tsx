import { AppearanceStyleType, Button, t } from "@bloom-housing/ui-components"
import React from "react"
import { Dialog } from "@bloom-housing/ui-seeds"

const AccountAlreadyConfirmedModal = ({
  isOpen,
  onClose,
}: {
  isOpen: boolean
  onClose: () => void
}) => {
  return (
    <Dialog isOpen={isOpen} onClose={onClose}>
      <Dialog.Header>{t("signIn.accountAlreadyConfirmed.title")}</Dialog.Header>
      <Dialog.Content>{t("signIn.accountAlreadyConfirmed.p1")}</Dialog.Content>
      <Dialog.Footer>
        <Button type="submit" styleType={AppearanceStyleType.primary} onClick={onClose}>
          {t("t.ok")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}

export { AccountAlreadyConfirmedModal as default, AccountAlreadyConfirmedModal }
