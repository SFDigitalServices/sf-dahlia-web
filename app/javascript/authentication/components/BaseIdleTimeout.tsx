import React, { FunctionComponent, useEffect, useState } from "react"

import {
  AlertTypes,
  AppearanceStyleType,
  Button,
  Modal,
  setSiteAlertMessage,
} from "@bloom-housing/ui-components"

const PROMPT_TIMEOUT = 60000
const events = ["mousemove", "keypress", "scroll"]

function useIdleTimeout(timeoutMs: number, onTimeout: () => void) {
  useEffect(() => {
    let timer: number = (setTimeout(onTimeout, timeoutMs) as unknown) as number
    const restartTimer = () => {
      if (timer) {
        clearTimeout(timer)
      }
      timer = (setTimeout(onTimeout, timeoutMs) as unknown) as number
    }

    // Listen for any activity events & reset the timer when they are found
    if (typeof document !== "undefined") {
      events.forEach((event) => document.addEventListener(event, restartTimer, false))
    }

    // Clean up our listeners & clear the timeout on unmounting/updating the effect
    return () => {
      if (timer) {
        clearTimeout(timer)
      }
      events.forEach((event) => document.removeEventListener(event, restartTimer, false))
    }
  }, [timeoutMs, onTimeout])
}

type IdleTimeoutProps = {
  promptTitle: string
  promptText: string
  promptAction: string
  redirectPath: string
  alertMessage: string
  alertType?: AlertTypes
  onTimeout: () => unknown
}

const BaseIdleTimeout: FunctionComponent<IdleTimeoutProps> = (props: IdleTimeoutProps) => {
  const {
    promptTitle,
    promptAction,
    promptText,
    redirectPath,
    alertMessage,
    alertType = "alert",
    onTimeout,
  } = props

  // 5 minutes
  const idleTimeout = 5 * 60 * 1000
  const [promptTimeout, setPromptTimeout] = useState<number | undefined>()

  useIdleTimeout(idleTimeout, () => {
    // Clear any existing prompt timeouts
    if (promptTimeout) {
      clearTimeout(promptTimeout)
    }

    const timeoutAction = async () => {
      setPromptTimeout(undefined)
      await onTimeout()
      setSiteAlertMessage(alertMessage, alertType)

      // replace this with proper react router when we have one
      return (window.location.href = redirectPath)
    }

    // Give the user 1 minute to respond to the prompt before the onTimeout action
    setPromptTimeout(
      (setTimeout(() => {
        void timeoutAction()
      }, PROMPT_TIMEOUT) as unknown) as number
    )
  })

  const modalActions = [
    <Button
      styleType={AppearanceStyleType.primary}
      onClick={() => {
        clearTimeout(promptTimeout)
        setPromptTimeout(undefined)
      }}
    >
      {promptAction}
    </Button>,
  ]

  return (
    <Modal
      open={Boolean(promptTimeout)}
      title={promptTitle}
      ariaDescription={promptText}
      actions={modalActions}
      hideCloseIcon
    >
      {promptText}
    </Modal>
  )
}

export default BaseIdleTimeout
