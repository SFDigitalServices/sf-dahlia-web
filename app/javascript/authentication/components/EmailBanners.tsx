import { t } from "@bloom-housing/ui-components"
import { Alert } from "@bloom-housing/ui-seeds"
import React from "react"
import { confirmEmail } from "../../api/authApiService"

export const requestEmail = (
  email: string,
  setEmailSent: (value: boolean) => void,
  setEmailSentError: (value: string | null) => void
) => {
  confirmEmail(email)
    .then(() => {
      setEmailSent(true)
    })
    .catch(() => {
      setEmailSentError(t("signIn.newAccount.sendEmailAgainButton.error"))
    })
}

export const EmailBanners = ({
  emailSent,
  emailSentError,
}: {
  emailSent: boolean
  emailSentError: string
}) => {
  return (
    <>
      {emailSent && (
        <Alert className="sign-in-banner banner-background-color">
          {t("signIn.newAccount.sendEmailAgainButton.confirmation")}
        </Alert>
      )}
      {emailSentError && (
        <Alert variant="alert" className="sign-in-banner">
          {emailSentError}
        </Alert>
      )}
    </>
  )
}
