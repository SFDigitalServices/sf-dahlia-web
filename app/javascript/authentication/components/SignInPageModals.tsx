import { AppearanceStyleType, Button, t } from "@bloom-housing/ui-components"
import React, { useState } from "react"
import { confirmEmail } from "../../api/authApiService"
import { Alert, Dialog } from "@bloom-housing/ui-seeds"

export const AccountAlreadyConfirmedModal = ({
  isOpen,
  onClose,
}: {
  isOpen: boolean
  onClose: () => void
}) => {
  return (
    <Dialog isOpen={isOpen} onClose={onClose}>
      <Dialog.Header>{t("signIn.accountAlreadyConfirmed.title")}</Dialog.Header>
      <Dialog.Content>{t("signIn.accountAlreadyConfirmed.p1")}</Dialog.Content>
      <Dialog.Footer>
        <Button type="submit" styleType={AppearanceStyleType.primary} onClick={onClose}>
          {t("t.ok")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}

export const NewAccountNotConfirmedModal = ({
  email,
  onClose,
}: {
  email: string
  onClose: () => void
}) => {
  const [emailSent, setEmailSent] = useState(false)
  const [emailSentError, setEmailSentError] = useState<string | null>(null)
  const requestEmail = () => {
    confirmEmail(email)
      .then(() => {
        setEmailSent(true)
      })
      .catch(() => {
        setEmailSentError(t("signIn.newAccount.sendEmailAgainButton.error"))
      })
  }

  return (
    <Dialog isOpen={!!email} onClose={onClose}>
      <Dialog.Header>{t("signIn.newAccount.title")}</Dialog.Header>
      {emailSent && (
        <Alert className="sign-in-banner banner-background-color">
          {t("signIn.newAccount.sendEmailAgainButton.confirmation")}
        </Alert>
      )}
      {emailSentError && (
        <Alert variant="alert" className="sign-in-banner">
          {emailSentError}
        </Alert>
      )}
      <Dialog.Content>{t("signIn.newAccount.p1", { email })}</Dialog.Content>
      <Dialog.Content>{t("signIn.newAccount.p2")}</Dialog.Content>
      <Dialog.Footer>
        <Button type="submit" styleType={AppearanceStyleType.primary} onClick={onClose}>
          {t("t.ok")}
        </Button>
        <Button styleType={AppearanceStyleType.secondary} onClick={requestEmail}>
          {t("signIn.newAccount.sendEmailAgainButton")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}
