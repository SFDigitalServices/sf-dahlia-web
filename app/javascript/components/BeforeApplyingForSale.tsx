import React from "react"
import { Heading, ListSection, t } from "@bloom-housing/ui-components"
import { renderMarkup } from "../util/languageUtil"

export enum BeforeApplyingType {
  DIRECTORY,
  LISTING_DETAILS,
  LISTING_DETAILS_HABITAT,
}
export interface BeforeApplyingForSaleProps {
  /**
   * Header will display differently depending on type
   */
  beforeApplyingType: BeforeApplyingType
}
export const BeforeApplyingForSale = ({ beforeApplyingType }: BeforeApplyingForSaleProps) => {
  const listClassNames = ["numbered-list", "my-5"]
  let listItems, requirements

  // extra styles on the details page
  if (
    beforeApplyingType === BeforeApplyingType.LISTING_DETAILS ||
    beforeApplyingType === BeforeApplyingType.LISTING_DETAILS_HABITAT
  ) {
    listClassNames.push("bg-white", "p-4", "pt-6", "text-xs")
  }

  // different requirements + list items for habitat vs other sales
  if (beforeApplyingType === BeforeApplyingType.LISTING_DETAILS_HABITAT) {
    requirements = renderMarkup(
      `${t("saleDirectory.beforeApplying.readFullList", {
        url: "https://habitatgsf.org/amber-drive-info/",
      })}`
    )

    listItems = [
      {
        content: renderMarkup(
          `${t("listingsForSale.beforeApplyingHabitat.step1", {
            url: "https://habitatgsf.org/amber-drive-info/",
          })}`
        ),
      },
      { content: t("listingsForSale.beforeApplyingHabitat.step2") },
      {
        content: renderMarkup(
          `${t("listingsForSale.beforeApplyingHabitat.step3", {
            url:
              "https://sfmohcd.org/sites/default/files/Documents/MOH/Inclusionary%20Manuals/Inclusionary%20Affordable%20Housing%20Program%20Manual%2010.15.2018.pdf",
          })}`
        ),
      },
      { content: t("listingsForSale.beforeApplyingHabitat.step4") },
    ]
  } else if (
    beforeApplyingType === BeforeApplyingType.DIRECTORY ||
    beforeApplyingType === BeforeApplyingType.LISTING_DETAILS
  ) {
    requirements = renderMarkup(
      `${t("saleDirectory.beforeApplying.readFullList", {
        url: "https://sfmohcd.org/homebuyer-program-eligibility",
      })}`
    )

    listItems = [
      { content: t("saleDirectory.beforeApplying.step1") },
      {
        content: renderMarkup(
          `${t("saleDirectory.beforeApplying.step2", {
            url: "https://sfmohcd.org/homebuyer-program-eligibility",
          })}`
        ),
      },
      {
        content: renderMarkup(
          `${t("saleDirectory.beforeApplying.step3", {
            url: "https://sfmohcd.org/homebuyer-application-requirements#education",
          })}`
        ),
      },
      {
        content: renderMarkup(
          `${t("saleDirectory.beforeApplying.step4", {
            url: "https://sfmohcd.org/lender-list",
          })}`
        ),
      },
      { content: t("saleDirectory.beforeApplying.step5") },
    ]
  }

  // similar format for list items, but they will be wrapped differently depending on if directory page or details page
  const content = (
    <>
      <ol className={[...listClassNames].join(" ")}>
        {listItems.map((item, index) => (
          <li className="text-gray-700" key={index}>
            {item.content}
          </li>
        ))}
      </ol>
      <div className="text-gray-750 primary-lighter-markup-link-desktop">{requirements}</div>
    </>
  )

  return beforeApplyingType === BeforeApplyingType.LISTING_DETAILS ||
    beforeApplyingType === BeforeApplyingType.LISTING_DETAILS_HABITAT ? (
    <ListSection
      title={t("saleDirectory.beforeApplying.title")}
      subtitle={t("saleDirectory.beforeApplying.makeSureYou")}
    >
      {content}
    </ListSection>
  ) : (
    <li className="mb-8 style-none">
      <Heading className="mb-5 text-gray-750" priority={2} styleType={"underlineWeighted"}>
        {t("saleDirectory.beforeApplying.title")}
      </Heading>
      <p className="text-gray-750">{t("saleDirectory.beforeApplying.makeSureYou")}</p>
      {content}
    </li>
  )
}
