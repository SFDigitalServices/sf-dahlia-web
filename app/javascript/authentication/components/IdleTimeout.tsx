import React, { useContext } from "react"

import { t } from "@bloom-housing/ui-components"

import { getHomepagePath, getSignInPath } from "../../util/routeUtil"
import UserContext from "../context/UserContext"
import BaseIdleTimeout from "./BaseIdleTimeout"

interface IdleTimeoutProps {
  onTimeout?: () => unknown
  useFormTimeout?: boolean
}

const IdleTimeout = ({ onTimeout, useFormTimeout = false }: IdleTimeoutProps) => {
  const { profile, timeOut } = useContext(UserContext)

  const handleTimeout = async (shouldTimeOut: boolean) => {
    onTimeout && (await onTimeout())
    timeOut && shouldTimeOut && timeOut()
  }

  // Only render the IdleTimeout component if the user is logged in
  if (profile && timeOut) {
    return (
      <BaseIdleTimeout
        promptTitle={t("idleTimeout.stayLoggedIn")}
        promptText={t("idleTimeout.sessionInactivityLoggedIn")}
        promptAction={t("t.continue")}
        redirectPath={getSignInPath()}
        alertMessage={t("signIn.userTokenValidationTimeout")}
        alertType={useFormTimeout ? "alert" : "notice"}
        onTimeout={() => handleTimeout(true)}
      />
    )
  }

  if (useFormTimeout) {
    return (
      <BaseIdleTimeout
        promptTitle={t("idleTimeout.stayLoggedIn")}
        promptText={t("idleTimeout.sessionInactivity")}
        promptAction={t("t.continue")}
        redirectPath={getHomepagePath()}
        alertMessage={t("idleTimeout.sessionExpired")}
        alertType={"alert"}
        onTimeout={() => handleTimeout(false)}
      />
    )
  }

  return null
}

export default IdleTimeout
