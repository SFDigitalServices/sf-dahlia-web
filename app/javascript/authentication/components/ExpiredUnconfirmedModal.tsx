import { t, Button, AppearanceStyleType } from "@bloom-housing/ui-components"
import { Dialog } from "@bloom-housing/ui-seeds"
import React, { useState } from "react"
import { EmailBanners, requestEmail } from "./EmailBanners"

export const ExpiredUnconfirmedModal = ({
  email,
  onClose,
}: {
  email: string
  onClose: () => void
}) => {
  const [emailSent, setEmailSent] = useState(false)
  const [emailSentError, setEmailSentError] = useState<string | null>(null)

  return (
    <Dialog isOpen={!!email} onClose={onClose}>
      <Dialog.Header>{t("signIn.expiredUnconfirmed.title")}</Dialog.Header>
      <EmailBanners emailSent={emailSent} emailSentError={emailSentError} />
      <Dialog.Content>{t("signIn.expiredUnconfirmed.p1")}</Dialog.Content>
      <Dialog.Content>{t("signIn.expiredUnconfirmed.p2")}</Dialog.Content>
      <Dialog.Footer>
        <Button
          type="submit"
          styleType={AppearanceStyleType.primary}
          onClick={() => requestEmail(email, setEmailSent, setEmailSentError)}
        >
          {t("signIn.expiredUnconfirmed.sendNewLink")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}
