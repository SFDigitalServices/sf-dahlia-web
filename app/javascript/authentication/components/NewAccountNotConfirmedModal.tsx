import { t, Button, AppearanceStyleType } from "@bloom-housing/ui-components"
import { Dialog } from "@bloom-housing/ui-seeds"
import React, { useState } from "react"
import { EmailBanners, requestEmail } from "./EmailBanners"

export const NewAccountNotConfirmedModal = ({
  email,
  onClose,
}: {
  email: string
  onClose: () => void
}) => {
  const [emailSent, setEmailSent] = useState(false)
  const [emailSentError, setEmailSentError] = useState<string | null>(null)

  return (
    <Dialog isOpen={!!email} onClose={onClose}>
      <Dialog.Header>{t("signIn.newAccount.title")}</Dialog.Header>
      <EmailBanners emailSent={emailSent} emailSentError={emailSentError} />
      <Dialog.Content>{t("signIn.newAccount.p1", { email })}</Dialog.Content>
      <Dialog.Content>{t("signIn.newAccount.p2")}</Dialog.Content>
      <Dialog.Footer>
        <Button type="submit" styleType={AppearanceStyleType.primary} onClick={onClose}>
          {t("t.ok")}
        </Button>
        <Button
          styleType={AppearanceStyleType.secondary}
          onClick={() => requestEmail(email, setEmailSent, setEmailSentError)}
        >
          {t("signIn.newAccount.sendEmailAgainButton")}
        </Button>
      </Dialog.Footer>
    </Dialog>
  )
}
