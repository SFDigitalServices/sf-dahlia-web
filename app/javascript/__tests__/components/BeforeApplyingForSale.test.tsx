import React from "react"
import renderer from "react-test-renderer"
import { BeforeApplyingForSale, BeforeApplyingType } from "../../components/BeforeApplyingForSale"

describe("BeforeApplyingForSale", () => {
  it("display Before Applying when type is directory", () => {
    const tree = renderer
      .create(<BeforeApplyingForSale beforeApplyingType={BeforeApplyingType.DIRECTORY} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("display Before Applying when type is listing detail", () => {
    const tree = renderer
      .create(<BeforeApplyingForSale beforeApplyingType={BeforeApplyingType.LISTING_DETAILS} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("display Before Applying when type is Habitat listing detail", () => {
    const tree = renderer
      .create(
        <BeforeApplyingForSale beforeApplyingType={BeforeApplyingType.LISTING_DETAILS_HABITAT} />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
