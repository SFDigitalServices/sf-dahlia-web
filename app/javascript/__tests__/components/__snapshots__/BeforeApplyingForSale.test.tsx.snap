// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeforeApplyingForSale display Before Applying when type is Habitat listing detail 1`] = `
<li
  className="list-section custom-counter__item"
>
  <header
    className="list-section__header custom-counter__header"
  >
    <hgroup>
      <h4
        className="custom-counter__title"
      >
        Before applying
      </h4>
      <span
        className="custom-counter__subtitle"
      >
        Make sure you:
      </span>
    </hgroup>
  </header>
  <ol
    className="numbered-list my-5 bg-white p-4 pt-6 text-sm"
  >
    <li>
      <p>
        Go to a 
        <a
          href="https://habitatgsf.org/amber-drive-info/"
          target="_blank"
        >
          Habitat for Humanity information session
        </a>
      </p>
    </li>
    <li>
      Haven't owned residential property in the past 3 years
    </li>
    <li>
      <p>
        Meet the 
        <a
          href="https://sfmohcd.org/sites/default/files/Documents/MOH/Inclusionary%20Manuals/Inclusionary%20Affordable%20Housing%20Program%20Manual%2010.15.2018.pdf"
          target="_blank"
        >
          income requirements
        </a>
      </p>
    </li>
    <li>
      Have a 650 credit score or higher
    </li>
  </ol>
  <p>
    Read 
    <a
      href="https://habitatgsf.org/amber-drive-info/"
      target="_blank"
    >
      our full list of requirements
    </a>
     for more details.
  </p>
</li>
`;

exports[`BeforeApplyingForSale display Before Applying when type is directory 1`] = `
<li
  className="mb-8 style-none"
>
  <h2
    className="text-caps-underline mb-5"
  >
    Before applying
  </h2>
  <p>
    Make sure you:
  </p>
  <ol
    className="numbered-list my-5"
  >
    <li>
      Haven't owned residential property in the past 3 years
    </li>
    <li>
      <p>
        Meet our 
        <a
          href="https://sfmohcd.org/homebuyer-program-eligibility"
          target="_blank"
        >
          income requirements
        </a>
      </p>
    </li>
    <li>
      <p>
        Complete 
        <a
          href="https://sfmohcd.org/homebuyer-application-requirements#education"
          target="_blank"
        >
          homebuyer education
        </a>
      </p>
    </li>
    <li>
      <p>
        Get pre-approved for a mortgage loan by a 
        <a
          href="https://sfmohcd.org/lender-list"
          target="_blank"
        >
          lender on our list
        </a>
      </p>
    </li>
    <li>
      Have enough in savings for closing costs and downpayment
    </li>
  </ol>
  <p>
    Read 
    <a
      href="https://sfmohcd.org/homebuyer-program-eligibility"
      target="_blank"
    >
      our full list of requirements
    </a>
     for more details.
  </p>
</li>
`;

exports[`BeforeApplyingForSale display Before Applying when type is listing detail 1`] = `
<li
  className="list-section custom-counter__item"
>
  <header
    className="list-section__header custom-counter__header"
  >
    <hgroup>
      <h4
        className="custom-counter__title"
      >
        Before applying
      </h4>
      <span
        className="custom-counter__subtitle"
      >
        Make sure you:
      </span>
    </hgroup>
  </header>
  <ol
    className="numbered-list my-5 bg-white p-4 pt-6 text-sm"
  >
    <li>
      Haven't owned residential property in the past 3 years
    </li>
    <li>
      <p>
        Meet our 
        <a
          href="https://sfmohcd.org/homebuyer-program-eligibility"
          target="_blank"
        >
          income requirements
        </a>
      </p>
    </li>
    <li>
      <p>
        Complete 
        <a
          href="https://sfmohcd.org/homebuyer-application-requirements#education"
          target="_blank"
        >
          homebuyer education
        </a>
      </p>
    </li>
    <li>
      <p>
        Get pre-approved for a mortgage loan by a 
        <a
          href="https://sfmohcd.org/lender-list"
          target="_blank"
        >
          lender on our list
        </a>
      </p>
    </li>
    <li>
      Have enough in savings for closing costs and downpayment
    </li>
  </ol>
  <p>
    Read 
    <a
      href="https://sfmohcd.org/homebuyer-program-eligibility"
      target="_blank"
    >
      our full list of requirements
    </a>
     for more details.
  </p>
</li>
`;
